### PR TITLE
[SYCL][CUDA][HIP] Support zero range kernel for cuda and hip backends.

### PR DESCRIPTION
**Pull request by mmoadeli**
- Fixes issue [6963](https://github.com/intel/llvm/issues/6963) to allow range zero kernel for cuda and hip backends.

### CODE REVIEW

File: sycl/plugins/cuda/pi_cuda.cpp
Comment (steffenlarsen) [inline sycl/plugins/cuda/pi_cuda.cpp:2613]: Doing an early exit here means that we do not create an event. I fear that could cause unexpected problems. Same goes for HIP.

**Review by steffenlarsen — COMMENTED** [review]
Thank you for submitting a fix for this, @mmoadeli! This definitely seems like a bug, given the [SYCL 2020 specification](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_work_group_data_parallel_kernels) states that:

> When the global size is zero, the kernel function is not executed, the local size is ignored, and any dependencies are satisfied.

I do however wonder if it would be better to handle this at the runtime level rather than in the individual plugins. Based on the new test I assume this is also currently not handled by the L0 plugin. @smaslov-intel - What do you think?

File: sycl/test/basic_tests/range_zero_size.cpp
Comment (steffenlarsen) [inline sycl/test/basic_tests/range_zero_size.cpp:18]: Device-dependent testing should be in https://github.com/intel/llvm-test-suite rather than in the in-tree LIT tests. Could you please move this test to there?

**Comment by romanovvlad** [comment]
> Thank you for submitting a fix for this, @mmoadeli! This definitely seems like a bug, given the [SYCL 2020 specification](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_work_group_data_parallel_kernels) states that:
> 
> > When the global size is zero, the kernel function is not executed, the local size is ignored, and any dependencies are satisfied.
> 
> I do however wonder if it would be better to handle this at the runtime level rather than in the individual plugins. Based on the new test I assume this is also currently not handled by the L0 plugin. @smaslov-intel - What do you think?

Probably it's not worth optimizing, but the SYCL RT can emulate the required behavior by submitting an RT barrier command only, while plugins can do something better.

File: sycl/test/basic_tests/range_zero_size.cpp
Comment (mmoadeli) [reply sycl/test/basic_tests/range_zero_size.cpp:18]: actually, if you agree, the device dependent part of the test to be removed. The same behaviour should be seen for all backends.

File: sycl/plugins/cuda/pi_cuda.cpp
Comment (mmoadeli) [reply sycl/plugins/cuda/pi_cuda.cpp:2613]: @steffenlarsen I agree with you on this regarding the event. 
If I  move the condition (**if (*global_work_size =! 0 )**) to guard the call to  **PI_CHECK_ERROR(cuLaunchKernel(** it will preserve the functionality related to event handling. Do you have any concern doing that?
I have not tried the L0, but the opencl:cpu and esimd (and host in an earlier version of DP++) backend are tried and work without any modifications.

Comment (steffenlarsen) [reply sycl/plugins/cuda/pi_cuda.cpp:2613]: Either that or you could copy the event creation to here, something like:

```c++
if (event) {
  std::unique_ptr<_pi_event> retImplEv{nullptr};
  retImplEv = std::unique_ptr<_pi_event>(_pi_event::make_native(
      PI_COMMAND_TYPE_NDRANGE_KERNEL, command_queue));
  retImplEv->start();
  retImplEv->record();
  *event = retImplEv.release();
}
```

Comment (romanovvlad) [reply sycl/plugins/cuda/pi_cuda.cpp:2613]: I'm not sure it will be a correct implementation. A call with `NDRange == 0` while should do nothing in terms of running the kernel, the event produced should be as if we run a kernel:
1. Completion of such an event should guarantee completion of events passed in `event_wait_list`
2. The event can be used in a `event_wait_list` of a subsequent enqueue.

So, I think that the more correct implementation would be just calling `cuda_piEnqueueEventsWaitWithBarrier` if `NDRange == 0`.

Comment (steffenlarsen) [reply sycl/plugins/cuda/pi_cuda.cpp:2613]: That is a good point. Now that there are multiple queues, recording an event is not enough to act like a barrier.

**Comment by mmoadeli** [comment]
@steffenlarsen I'd be happy to address any potential issues on this PR to have it merged.

**Review by steffenlarsen — APPROVED** [review]
Sorry for the delay! Changes look good. Where did you move the tests to?

**Comment by mmoadeli** [comment]
> Sorry for the delay! Changes look good. Where did you move the tests to?
I had a test here, but I was advised to move it to llvm-tests-suite. I was waiting for this ones approval and potential merge to add the test to llvm-tests-suies

**Comment by steffenlarsen** [comment]
> > Sorry for the delay! Changes look good. Where did you move the tests to?
> > @steffenlarsen  I had a test here, but I was advised to move it to llvm-tests-suite. I was waiting for this ones approval and potential merge to add the test to llvm-tests-suies
> 
> The test can be found in first commit [719c6b8](https://github.com/intel/llvm/commit/719c6b894b8f88c512479e699d76ec5530f5b599)

It would be preferable if you can open a PR on the test-suite with the test. We can trigger testing with it on this PR, albeit only on L0 and OCL.

**Comment by mmoadeli** [comment]
> 
@steffenlarsen [Test Suite PR 1363](https://github.com/intel/llvm-test-suite/pull/1363)

**Comment by steffenlarsen** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/1363

File: sycl/plugins/cuda/pi_cuda.cpp
Comment (romanovvlad) [inline sycl/plugins/cuda/pi_cuda.cpp:2994]:
```suggestion
    return cuda_piEnqueueEventsWaitWithBarrier(command_queue, num_events_in_wait_list,
                                        event_wait_list, event);
```

File: sycl/plugins/hip/pi_hip.cpp
Comment (romanovvlad) [inline sycl/plugins/hip/pi_hip.cpp:2764]:
```suggestion
    return hip_piEnqueueEventsWaitWithBarrier(command_queue, num_events_in_wait_list,
                                       event_wait_list, event);
```

**Review by romanovvlad — APPROVED** [review]

**Pull request merged**
